GitUI/BranchTreePanel/RepoObjectsTree.cs

```diff
@@ -491,6 +492,58 @@ List<TreeNode> SearchTree(string text, IEnumerable<TreeNode> nodes)
                     }
                 }
 
+                // if matches with high priority are found bring those to the first of the list by rotating the list
+                // Rotation is done to maintain the tree order, so that on click of search button multiple times,
+                // the selected node passes down the tree to other matches.
+                if (matchIndex > 0)
+                {
+                    rotateListToSetFirstAsMatchIndex(ret, matchIndex);
+                }
+
+                // A function to see if the node text or full path of a node (for branch node) is matching with search text
+                bool searchForMatchingText(string nodeText, string searchText, ref bool isFullMatch, ref int matchIndex, List<TreeNode> ret)
+                {
+                    if (nodeText.IndexOf(searchText, StringComparison.InvariantCultureIgnoreCase) != -1)
+                    {
+                        // If full match is already found no need to search for furthur matches as full match has highest priority
+                        if (!isFullMatch)
+                        {
+                            // checking for full match
+                            if (string.Compare(nodeText, searchText, StringComparison.InvariantCultureIgnoreCase) == 0)
+                            {
+                                isFullMatch = true;
+                                matchIndex = ret.Count;
+                            }
+
+                            // if full match not found, check for matches appearing after the last '/' for child node match.
+                            // If a match after '/' was already found then skip this by checking match index
+                            else if (matchIndex == -1 && nodeText.Contains('/') &&
+                                string.Compare(searchText, nodeText.Split('/').Last(), StringComparison.InvariantCultureIgnoreCase) == 0)
+                            {
+                                matchIndex = ret.Count;
+                            }
+                        }
+
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                // Rotate the list to set the match with high priority as first item
+                void rotateListToSetFirstAsMatchIndex(List<TreeNode> ret, int matchIndex)
+                {
+                    int rotationIndex = 0;
+
+                    while (rotationIndex < matchIndex)
+                    {
+                        TreeNode firstElement = ret.FirstOrDefault();
+                        ret.RemoveAt(0);
+                        ret.Add(firstElement);
+                        rotationIndex++;
+                    }
```

Comment:
This *is* a classical `for` loop - no need for ``while`.
```suggestion
                    for(int rotationIndex = 0; rotationIndex < matchIndex; ++rotationIndex)
                    {
                        TreeNode firstElement = ret.FirstOrDefault();
                        ret.RemoveAt(0);
                        ret.Add(firstElement);
                    }
```

Reply:
Yes, I had some other logic for rotation in beginning, which required while loop. Missed to change it to for loop.
Thank you.
How do I proceed with these commit suggestions? Do I commit it from here or should I make change in my local and push to fork? (This is my first PR so trying to understand the flow)

Reply:
The normal is to make fixup commits until approved.
In this case it is OK to squash and rebase on master

Reply:
I had made a merge from remote/branch (5ce4cbc) after my first commit. Is it okay to squash this also?

Reply:
> I had made a merge from remote/branch ([5ce4cbc](https://github.com/gitextensions/gitextensions/commit/5ce4cbc6e37a2eb755a47de79135e6bf0ccabd8f)) after my first commit. Is it okay to squash this also?

Yes, please. We use to rebase in order to avoid all merges except PR to master - (and even those if supported by GH).

Reply:
I just edited my comment above on how to do the rebase in this scenario. Can you please give your inputs on the same.

Reply:
There are many ways to rebase. Some info in the wiki https://github.com/gitextensions/gitextensions/wiki/How-To%3A-Squash-and-Rebase-your-changes